### DISCUSSION TITLE
How can we discover the value to add to system.properties libpath when a native lib is missing after app signing

### DISCUSSION THREAD

**Discussion by csandels** [hydraulic-software/conveyor#90] [Q&A] [answered]
Our compose Multiplatform app uses the mongo realm sdk and skiko libraries (for example).  It runs perfectly when it's not signed.  Once it's signed, it fails to run and logs errors regarding inability to load native libraries.

Conveyor addresses this issue in the Troubleshooting page at https://conveyor.hydraulic.dev/12.1/troubleshooting/troubleshooting-jvm/ under the section "Problems loading native libraries".  The suggestion is to add the native libs to the config like: 
```
app {
    jvm {
        system-properties {
            "my.library.nativePathLib" = <libpath>
        }
    }
}
```
But how do we know what the value of "my.library.nativePathLib" should actually be for each lib?  Is it a class path or something from a manifest somewhere or a gradle file for the lib?  Looks like `<libpath>` is a literal value but assuming nativePathLib is not.  For example, for skiko, would we use the following?:

`"libskiko" = <libpath>`

**Comment by csandels** [comment]
Error given when attempting to run the app which is signed by conveyor.  We don't get this error when building and signing the app with the standard gradle toolchain.  Also, the librealmc.dylib file does exist at the path given in the trace.

java.lang.ExceptionInInitializerError
at io.realm.kotlin.internal.interop.RealmInterop.realm_network_transport_new(RealmInterop.kt:1266)
at io.realm.kotlin.mongodb.internal.AppConfigurationImpl.initializeRealmAppConfig(AppConfigurationImpl.kt:130)
at io.realm.kotlin.mongodb.internal.AppConfigurationImpl.createNativeApp(AppConfigurationImpl.kt:76)
at io.realm.kotlin.mongodb.internal.AppImpl.<init>(AppImpl.kt:97)
at io.realm.kotlin.mongodb.App$Companion.create(App.kt:152)
...
Caused by: java.lang.RuntimeException: Couldn't load Realm native libraries
at io.realm.kotlin.internal.interop.realmc.<clinit>(realmc.java:28)
... 22 more
Caused by: java.lang.reflect.InvocationTargetException
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:568)
at io.realm.kotlin.internal.interop.realmc.<clinit>(realmc.java:26)
... 22 more
Caused by: java.lang.UnsatisfiedLinkError: Can't load library: /Users/abc/Library/Caches/io.realm.kotlin/cc4e9f010ff66ddb54224ecd4aca9c6a44f114ca/librealmc.dylib
at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2393)
at java.base/java.lang.Runtime.load0(Runtime.java:755)
at java.base/java.lang.System.load(System.java:1953)
at io.realm.kotlin.jvm.SoLoader.load(SoLoader.kt:83)
at io.realm.kotlin.jvm.SoLoader.load(SoLoader.kt:60)
... 27 more

**Comment by csandels** [comment]
Strangely, this issue is resolved when we have the following line in the conveyor.conf file.
`include required("https://raw.githubusercontent.com/hydraulic-software/conveyor/master/configs/jvm/extract-native-libraries.conf")`

We considered this before but didn't see anything about realm in the included file so figured it wouldn't help in this case.  Why does this line fix our issue when nothing about realm is present in the file?

**Reply by mikehearn** [reply, marked as answer]
There are two ways to handle JVM libraries that have native components:

1. Let them keep their shared libs inside JARs. In this case they will tend to extract the libraries to some per-library cache location in the user's home directory.
2. Extract the libraries ahead of time, at packaging time, and put them in the right place along with other native libraries that come with Java.

(2) is better for the end user. It means apps start faster, don't waste disk space, and don't clutter the home directory. However because native code handling on the JVM is kind of a mess, some libraries don't follow best practices and need to be explicitly configured to tell them not to extract their library. Usually via some system property, but the names aren't consistent so we can't tell you what it is. Others do the right thing and don't have to be configured.

We originally defaulted to (2), but too many developers found it confusing to have to configure , so we changed the default to (1). This tends to make apps 'just work'. Unfortunately on macOS there is an exception to this expectation, which is that _for developers only_ they can end up with two different versions of the same library, one signed and one not (or one signed by a different identity). macOS doesn't like mixing identities by default and will reject the load request, Java then swallows the error message and it's all kind of a disaster.

One fix is to just delete the cached library. When you start the app from the packaged form, it'll re-extract it but this time it'll be the correctly signed version and things should work. The other is to turn on native library extraction, which is what you've done by adding that include line. Then the realm library will be placed inside the app bundle in the expected location, and no runtime caching stuff is required.

**Reply by csandels** [reply]
And just for anyone else who may read this.  In #2, "the right place", in the case of Mac, means in the app Contents folder of the distributable app itself.  When we added the include line, that added `extract-native-libraries = true`.  Conveyor then signed the native realm lib and put it in the lib folder in the app package with the other native libs.  The JVM loader knows to look there and the realm code loading the native lib tries to get it on the standard system loader first, which finds it there if the include was used.  Thanks Mike.
